qemu/tests/aio_test.py

```diff
@@ -0,0 +1,126 @@
+    paths = os.environ.get("PATH", "").split(":")
+    for path in paths:
+        cmd_path = os.path.join(path, cmd)
+        if os.path.isfile(cmd_path) and os.access(cmd_path, os.X_OK):
+            return cmd_path
+    raise TestCancel("No command %s available" % cmd)
+
+
+def coroutine(func):
+    """Start coroutine."""
+    @functools.wraps(func)
+    def start(*args, **kargs):
+        cr = func(*args, **kargs)
+        cr.send(None)
+        return cr
+    return start
+
+
+def get_qemu_version(target=None):
+    """Get installed QEMU version."""
+    logging.debug("check QEMU version")
+    line = process.run("qemu-img --version").stdout_text.splitlines()[0]
+    version = line.split()[-1].strip("()")
+    logging.debug("QEMU version: %s", version)
+    try:
+        target.send(version)
+    except StopIteration:
+        pass
+
+
+@coroutine
+def brew_download_build(target):
+    """Download source rpm."""
+    version = yield
+    filename = "%s.src.rpm" % version
+    root_dir = data_dir.get_data_dir()
+    save_path = os.path.join(root_dir, filename)
+    logging.debug("download source rpm to %s", save_path)
+    if not os.path.isfile(save_path):
+        cwd = os.getcwd()
+        os.chdir(root_dir)
+        cmd = "brew download-build -q --rpm {filename}".format(
+            filename=filename)
+        try:
+            process.run(cmd)
+        finally:
+            os.chdir(cwd)
+    target.send(save_path)
+
+
+@coroutine
+def unpack_source(target):
+    """Unpack source rpm."""
+    path = yield
+    logging.debug("unpack source rpm")
+    process.run("rpm -ivhf {path}".format(path=path))
+    process.run("rpmbuild -bp /root/rpmbuild/SPECS/qemu-kvm.spec --nodeps")
+    version = re.search(r"\d+.\d+.\d+", path).group()
+    src_path = glob.glob("/root/rpmbuild/BUILD/qemu*%s" % version)[0]
+    target.send(src_path)
+
+
+@coroutine
+def run_aio_tests(target):
+    """Compile the source code then run aio tests."""
+    path = yield
+    cwd = os.getcwd()
+    os.chdir(path)
+    try:
+        logging.debug("compile source code of QEMU")
+        process.run("./configure")
+        cpu_count = cpu.online_cpus_count()
+        aio_path = "tests/test-aio"
+        make_cmd = "make {aio_path} -j{cpu_count}".format(
+            aio_path=aio_path, cpu_count=cpu_count)
+        process.run(make_cmd)
+        logging.debug("run aio tests")
+        result = process.run(aio_path)
+    finally:
+        os.chdir(cwd)
+    target.send(result.stdout_text)
+
+
+@coroutine
+def parse_result():
+    """Parse tests result."""
+    result = yield
+    err = False
+    for line in result.splitlines():
+        if "OK" not in line:
+            err = True
+            logging.error(line)
+    if err:
+        raise TestFail("aio test failed.")
+    else:
+        logging.debug("all aio tests have passed")
+
+
+def run(test, params, env):
+    """Build and run aio tests from QEMU source."""
```

Comment:
"""
    Build and run aio tests from QEMU source.

    Test steps:
    1). get qemu version
    2). brew download and build
    3). make test-aio
    4). run the aio tests
    """